lib/api/controllers/bulkController.js

```diff
@@ -53,47 +44,71 @@ class BulkController extends BaseController {
    * @param {Request} request
    * @returns {Promise}
    */
-  import(request) {
-    assertHasBody(request);
-    assertBodyHasAttribute(request, 'bulkData');
-
-    return this.engine.import(request)
-      .then(response => {
-        if (response.partialErrors && response.partialErrors.length > 0) {
+  import (request) {
+    const
+      userId = this.getUserId(request),
+      { index, collection } = this.getIndexAndCollection(request),
+      refresh = this.getStringParam(request, 'args.refresh', 'false'),
+      bulkData = this.getArrayParam(request, 'body.bulkData'),
+      options = {
+        refresh,
+        userId
+      };
+
+    this.assertIndexAndCollectionExists(index, collection);
+
+    return this.storageEngine.import(index, collection, bulkData, options)
+      .then(({ result, errors }) => {
+        if (errors.length > 0) {
           request.setError(errorsManager.getError(
             'document_creations_failed',
-            response.partialErrors));
+            errors));
         }
 
-        return response;
+        const items = result.concat(errors);
+
+        return {
+          items,
+          errors: errors.length > 0
+        };
       });
   }
 
   /**
    * Write a document without adding metadata or performing data validation.
+   *
    * @param {Request} request
    * @returns {Promise<Object>}
    */
-  write(request) {
-    assertHasBody(request);
-    assertHasIndexAndCollection(request);
-    assertIdStartsNotUnderscore(request);
-
-    const notify = this.tryGetBoolean(request, 'args.notify');
-
-    return this.engine.createOrReplace(request, false)
-      .then(response => {
-        this.kuzzle.indexCache.add(
-          request.input.resource.index,
-          request.input.resource.collection);
-
-        if (notify && response.created) {
-          this.kuzzle.notifier.notifyDocumentCreate(request, response);
+  write (request) {
+    const
+      { index, collection } = this.getIndexAndCollection(request),
+      id = this.getId(request),
+      content = this.getObjectParam(request, 'body'),
+      refresh = this.getStringParam(request, 'args.refresh', 'false'),
+      notify = this.tryGetBoolean(request, 'args.notify');
+
+    this.assertIndexAndCollectionExists(index, collection);
+
+    return this.storageEngine.createOrReplace(
+      index,
+      collection,
+      id,
+      content,
+      { refresh, injectKuzzleMeta: false }
+    )
+      .then(result => {
+        if (notify && result.created) {
+          this.kuzzle.notifier.notifyDocumentCreate(request, result);
         } else if (notify) {
           this.kuzzle.notifier.notifyDocumentReplace(request);
         }
 
-        return response;
+        return {
+          _id: result._id,
+          _version: result._version,
+          _source: result._source,
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this trailing comma. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3ATrailingComma)